### PR TITLE
the final split

### PR DESCRIPTION
**Pull request by debris**
@chfast @arkpar @LefterisJP @subtly @gluk256 please build and let me now if it is working or not. @gavofyork please verify the structure of directories

you can build everything together or submodules separately.

available options:

```cmake
	# features
	eth_default_option(VMTRACE OFF)
	eth_default_option(PROFILING OFF)
	eth_default_option(FATDB ON)
	eth_default_option(ROCKSDB OFF)
	eth_default_option(OLYMPIC OFF)
	eth_default_option(PARANOID OFF)

	# components
	eth_default_option(GUI ON)
	eth_default_option(TESTS ON)
	eth_default_option(TOOLS ON)
	eth_default_option(ETHASHCL ON)
	eth_default_option(EVMJIT OFF)
	eth_default_option(SOLIDITY ON)
```

changes in submodules:
- https://github.com/ethereum/alethzero/pull/39
- https://github.com/ethereum/mix/pull/49
- https://github.com/ethereum/webthree/pull/27
- https://github.com/debris/dev/commit/1618316c3efc48d3d63ebe63b8fa859bd0cf064a
- https://github.com/debris/ethereum/commit/e7db4c4563b5df47e54a754b020cf0bd47696589

### CODE REVIEW

**Comment by LefterisJP** [comment]
temporarily disabling the build options won't let me build. I need to disable EVMJIT in my system until the llvm 3.7 package is officially released.

**Comment by chfast** [comment]
I don't know what have been removed, but submodule build works after updating submodules.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,479 +1,44 @@
-# cmake global
 cmake_minimum_required(VERSION 3.0.0)
 
-# Figure out environment.
-set(ETH_CMAKE_DIR   "${CMAKE_CURRENT_LIST_DIR}/../cpp-ethereum-cmake"   CACHE PATH "The path to the cmake directory")
+set(ETH_CMAKE_DIR "${CMAKE_CURRENT_LIST_DIR}/cmake" CACHE PATH "The path to the cmake directory")
```
Comment (chfast) [inline CMakeLists.txt]: I think `"${CMAKE_SOURCE_DIR}/cmake"` would be much clearer to express the submodules are direct subdirectories.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,479 +1,44 @@
...
-message("-- GUI              Build GUI components                     ${GUI}")
-message("-- TESTS            Build tests                              ${TESTS}")
-message("-- ETHASHCL         Build OpenCL components                  ${ETHASHCL}")
-message("-- JSCONSOLE        Build with javascript console            ${JSCONSOLE}")
-message("-- EVMJIT           Build LLVM-based JIT EVM                 ${EVMJIT}")
-message("------------------------------------------------------------------------")
-message("")
-
-set(CMAKE_THREAD_LIBS_INIT pthread)
+project(cpp-ethereum VERSION "0.9.91")
```
Comment (chfast) [inline CMakeLists.txt]: project() sets variables like `<PROJECT-NAME>_VERSION`. I'm not sure it will be able to set `cpp-ethereum_VERSION` variable.

What is the meaning of the version any way?

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,479 +1,44 @@
...
-find_program(GRAPHVIZDOT dot)
-if (GRAPHVIZDOT)
-	message(STATUS "Graphviz detected. Drawing a module dependency graph at ${CMAKE_SOURCE_DIR}/dependency_graph.svg")
-	execute_process(
-		COMMAND dot -Tsvg dependency_graph.dot -o dependency_graph.svg
-		WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
-	)
+if (GUI)
+	add_subdirectory(alethzero)
+	add_subdirectory(mix)
```
Comment (chfast) [inline CMakeLists.txt]: Allow users not to clone optional submodules.

File: OldCMakeLists.cmake [hunk 4]
```diff
@@ -0,0 +1,246 @@
+# cmake global
```
Comment (chfast) [inline OldCMakeLists.cmake]: Is this file needed?

File: CMakeLists.txt (hunk 2 above)
Comment (debris) [inline CMakeLists.txt]: I've checked that and `cpp-ethereum_VERSION` variable is properly set.

> What is the meaning of the version any way?

what do you mean?

File: CMakeLists.txt (hunk 1 above)
Comment (debris) [inline CMakeLists.txt]: I stopped using `CMAKE_SOURCE_DIR`, cause it always points to top most CMakeLists.txt directory. I find using `CMAKE_CURRENT_LIST_DIR` better, cause it always points to the current cmake file directory, no matter if it is included by other cmake files.

File: CMakeLists.txt (hunk 3 above)
Comment (debris) [inline CMakeLists.txt]: They are allowed. If the run `cmake -DGUI=0` alethzero and mix submodules can be still uninitalized.

File: OldCMakeLists.cmake (hunk 4 above)
Comment (debris) [inline OldCMakeLists.cmake]: no. I will remove it

File: CMakeLists.txt (hunk 2 above)
Comment (chfast) [inline CMakeLists.txt]: That's for checking that. Have you checked also if the old `ethereum_VERSION` and `ethereum_SOURCE_DIR` have been replaced also?

It's umbrella project. It's optional. Does anyone care about its version? Does libethereum and AZ have theirs own versions? Should you bump it depending on what?

Maybe it's reasonable to remove the version here.

Comment (gavofyork) [inline CMakeLists.txt]: indeed - cpp-ethereum shouldn't actually contain anything substantial so doesn't need a (user-readable) version. commit hash will be fine.
